FLIP 245: Cadence Date and time (#245)
onflow/flips: cadence/20240115-cadence-date-time.md

---
status: draft
flip: 245
authors: darkdrag00n (darkdrag00n@proton.me)

> sponsor: Bastian Müller (bastian@dapperlabs.com)

Reviewer: ```suggestion
sponsor: Bastian Müller (bastian.mueller@flowfoundation.org)
```

updated: 2024-01-29
---

# FLIP 245: Date & Time

## Objective

This FLIP proposes adding Date & Time types and manipulation utilities to Cadence. It proposes two new types, `LocalDateTime` and `Duration` along with functions to create and manipulate them.

## Motivation

Manipulating Date & Time is a common aspect of user programs. Presently, Cadence users have to implement these calculations using raw `UFix64` timestamps.

Doing them manually is both hard and error-prone. As a result, standard library support similar to other modern languages would enhance the usability of the language.

## User Benefit

Manipulating date and time using raw timestamps is difficult and error-prone. Providing rich suite of standard library types and functions would increase the usability and user-experience.

## Design Proposal

> New types `LocalDateTime` & `Duration` will be added to Cadence. Both of these data types assume that there are 3600*24 seconds in every day and there are no leap years.

Reviewer: I wonder if we should add this functionality as a standard library contract, i.e. namespace it and not add them to the global namespace.

### LocalDateTime

The `LocalDateTime` type will be defined as follows:

```cadence

> pub struct LocalDateTime {

Reviewer: The proposal needs to be updated to Cadence 1.0. For example:
- `pub` -> `access(all)`
- Add `view` function modifier to getter functions

2. The type will **not** be timezone aware i.e. it'll be the responsibility of the application developer to handle timezone details.
3. It is assumed that there are 3600*24 seconds in every day and there are no leap years.

Apart from the `LocalDateTime`, a few other types will be added to Cadence. They are:

1. `Month`
2. `Day`

```cadence
pub enum Month: UInt8 {
    pub case january
    pub case february
    pub case march
    pub case april
    pub case may
    pub case june
    pub case july
    pub case august
    pub case september
    pub case october
    pub case november
    pub case december
}

pub enum Day: UInt8 {
    pub case monday
    pub case tuesday
    pub case wednesday
    pub case thursday
    pub case friday
    pub case saturday
    pub case sunday
}
```

#### Constructor Functions
Constructor functions will be defined to instantiate variables of type `LocalDateTime`. They are defined below.

- `LocalDateTime.now()`: Creates a `LocalDateTime` with value based on the current block timestamp.

- `LocalDateTime.fromTimestamp(timestamp : UFix64)`: Creates a `LocalDateTime` with value based on the passed timestamp argument.

> - `LocalDateTime.of(year: UInt64, month: Month, day: UInt8, hour: UInt8, minute: UInt8, second: UInt8)`: Creates a `LocalDateTime` with value based on the passed arguments assuming that there are 3600*24 seconds in every day.

Reviewer: This can maybe just be a constructor function. It might be good to improve the naming of `day` and make it explicit that it is the day of the month, e.g. `dayOfMonth`:

```suggestion
- `LocalDateTime(year: UInt64, month: Month, dayOfMonth: UInt8, hour: UInt8, minute: UInt8, second: UInt8)`: Creates a `LocalDateTime` with value based on the passed arguments assuming that there are 3600*24 seconds in every day.
```

Examples below also need to be updated

#### Public Members
The `LocalDateTime` will provide the following public members:

1. `getYear(): UInt32`: Get the year of the date-time.
2. `getMonth(): Month`: Get the month of the date-time.

> 3. `getDate(): UInt8`: Get the date of the date-time.

Reviewer: Maybe improve the naming and clarify it is the day of the month, i.e.

```suggestion
3. `getDayOfMonth(): UInt8`: Get the day-of-month of the date-time.
```

Examples below also need to be updated

4. `getHour(): UInt8`: Get the hour of the date-time.
5. `getMinute(): UInt8`: Get the minute of the date-time.
6. `getSecond(): UInt8`: Get the second of the date-time.
7. `getDayOfWeek(): Day`: Get the day-of-week of the date-time.

#### Examples
Example usage of `LocalDateTime` are:

```cadence
/// Assume that the current block timestamp is 1707045210
let currentDateTime = LocalDateTime.now()
currentDateTime.getYear()      // 2024
currentDateTime.getMonth()     // february
currentDateTime.getDate()      // 04
currentDateTime.getHour()      // 11
currentDateTime.getMinute()    // 13
currentDateTime.getSecond()    // 30
currentDateTime.getDayOfWeek() // sunday

let dateTimeFromTs = LocalDateTime.fromTimestamp(1415829132)
dateTimeFromTs.getYear()      // 2014
dateTimeFromTs.getMonth()     // november
dateTimeFromTs.getDate()      // 12
dateTimeFromTs.getHour()      // 21
dateTimeFromTs.getMinute()    // 52
dateTimeFromTs.getSecond()    // 12
dateTimeFromTs.getDayOfWeek() // wednesday

> let dateTimeFromComponents = LocalDateTime.of(2019, april, 29, 19, 49, 31)

Reviewer: Need to add argument labels, as required by the function declaration

3. `-999999999 <= days <= 999999999`

This model is taken from the time-tested Python [datetime module](https://docs.python.org/3/library/datetime.html#timedelta-objects).

#### Constructor Functions
Constructor functions will be defined to instantiate variables of type `Duration`. They are defined below.

- `Duration.betweenLocalDateTime(t1 : LocalDateTime, t2 : LocalDateTime)`: Creates a `Duration` which represents the time duration between arguments `t1` and `t2`.

- `Duration.of(years: Int64, days: Int64, hours: Int64, minutes: Int64, seconds: Int64, miliseconds: Int64, microseconds: Int64)`: Creates a `Duration` with value based on the passed arguments assuming that there are 3600*24 seconds in every day.

During the normalization:
1. 1 year is converted to 365 days
2. 1 hour is converted to 3600 seconds
3. 1 minute is converted to 60 seconds
4. 1 milisecond is converted to 1000 microseconds

#### Public Members
The `Duration` will provide the following public members:

- `getDays(): Int32`: Get the days of the `Duration`.
- `getSeconds(): UInt32`: Get the second of the `Duration`.
- `getMicroseconds(): UInt32`: Get the microseconds of the `Duration`.

> - `addYears(years : UInt32): Duration`: Return a new `Duration` after adding the provided number of years to it.

Reviewer: Here and for the other functions below: The function name already clarifies what the parameter is, so the argument label can be removed:

```suggestion
- `addYears(_ years: UInt32): Duration`: Return a new `Duration` after adding the provided number of years to it.
```

> - `addYears(years : UInt32): Duration`: Return a new `Duration` after adding the provided number of years to it.

Reviewer: The `add` name sounds like it is mutating, but it seems like that isn't the case right?

Maybe we can use the `plus` and `minus` naming like in Java?
